File: usedesk_sdk/src/main/java/ru/usedesk/sdk/external/entity/exceptions/UsedeskSocketException.java
package ru.usedesk.sdk.external.entity.exceptions;

import android.support.annotation.NonNull;

import static ru.usedesk.sdk.external.entity.exceptions.UsedeskSocketException.Error.UNKNOWN_ERROR;

public class UsedeskSocketException extends UsedeskException {
    private final Error error;

    public UsedeskSocketException() {
        this(UNKNOWN_ERROR);
    }

    public UsedeskSocketException(String message) {
        this(UNKNOWN_ERROR, message);
    }

    public UsedeskSocketException(@NonNull Error error) {
        this.error = error;
    }

    public UsedeskSocketException(@NonNull Error error, String message) {
        super(message);
        this.error = error;
    }

    @NonNull
    public Error getError() {
        return error;
    }

    public enum Error {
        INTERNAL_SERVER_ERROR,
        BAD_REQUEST_ERROR,
        FORBIDDEN_ERROR,
        DISCONNECTED,
        IO_ERROR,
        JSON_ERROR,
        UNKNOWN_FROM_SERVER_ERROR,
        UNKNOWN_ERROR
    }
}

File: common-sdk/src/main/java/ru/usedesk/common_sdk/external/entity/exceptions/UsedeskSocketException.kt
package ru.usedesk.common_sdk.external.entity.exceptions

class UsedeskSocketException : UsedeskException {
    val error: Error

    constructor(message: String?) : this(Error.UNKNOWN_ERROR, message)

    @JvmOverloads
    constructor(error: Error = Error.UNKNOWN_ERROR) {
        this.error = error
    }

    constructor(error: Error, message: String?) : super(message) {
        this.error = error
    }

    override fun toString() = super.toString() + "\n" + error.toString()

    enum class Error {
        INTERNAL_SERVER_ERROR,
        BAD_REQUEST_ERROR,
        FORBIDDEN_ERROR,
        DISCONNECTED,
        IO_ERROR,
        JSON_ERROR,
        UNKNOWN_FROM_SERVER_ERROR,
        UNKNOWN_ERROR
    }
}